// ternoa-js/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.chainConstants = exports.chainQuery = exports.txEvent = exports.txActions = exports.txPallets = void 0;
var txPallets;
(function (txPallets) {
    txPallets["marketplace"] = "marketplace";
    txPallets["nfts"] = "nfts";
    txPallets["utility"] = "utility";
    txPallets["balances"] = "balances";
    txPallets["capsules"] = "capsules";
    txPallets["associatedAccounts"] = "associatedAccounts";
    txPallets["system"] = "system";
})(txPallets = exports.txPallets || (exports.txPallets = {}));
var txActions;
(function (txActions) {
    txActions["buy"] = "buy";
    txActions["list"] = "list";
    txActions["unlist"] = "unlist";
    txActions["burn"] = "burn";
    txActions["create"] = "create";
    txActions["transfer"] = "transfer";
    txActions["transferAll"] = "transferAll";
    txActions["transferKeepAlive"] = "transferKeepAlive";
    txActions["finishSeries"] = "finishSeries";
    txActions["batch"] = "batch";
    txActions["batchAll"] = "batchAll";
    txActions["createFromNft"] = "createFromNft";
    txActions["remove"] = "remove";
    txActions["setIpfsReference"] = "setIpfsReference";
    txActions["setAltvrUsername"] = "setAltvrUsername";
    txActions["setCommissionFee"] = "setCommissionFee";
    txActions["setOwner"] = "setOwner";
    txActions["setKind"] = "setMarketType";
    txActions["setName"] = "setName";
    txActions["setUri"] = "setUri";
    txActions["setLogoUri"] = "setLogoUri";
})(txActions = exports.txActions || (exports.txActions = {}));
var txEvent;
(function (txEvent) {
    txEvent["nftsCreated"] = "Created";
    txEvent["nftsBurned"] = "Burned";
    txEvent["nftsTransfered"] = "Transfered";
    txEvent["CapsuleIpfsReferenceChanged"] = "CapsuleIpfsReferenceChanged";
    txEvent["CapsuleCreated"] = "CapsuleCreated";
    txEvent["CapsuleRemoved"] = "CapsuleRemoved";
    txEvent["MarketplaceCreated"] = "MarketplaceCreated";
    txEvent["ExtrinsicSuccess"] = "ExtrinsicSuccess";
    txEvent["ExtrinsicFailed"] = "ExtrinsicFailed";
    txEvent["BatchCompleted"] = "BatchCompleted";
    txEvent["BatchInterrupted"] = "BatchInterrupted";
})(txEvent = exports.txEvent || (exports.txEvent = {}));
var chainQuery;
(function (chainQuery) {
    chainQuery["nftMintFee"] = "nftMintFee";
    chainQuery["capsuleMintFee"] = "capsuleMintFee";
    chainQuery["marketplaceMintFee"] = "marketplaceMintFee";
    chainQuery["account"] = "account";
    chainQuery["number"] = "number";
})(chainQuery = exports.chainQuery || (exports.chainQuery = {}));
var chainConstants;
(function (chainConstants) {
    chainConstants["existentialDeposit"] = "existentialDeposit";
})(chainConstants = exports.chainConstants || (exports.chainConstants = {}));
